master/internal/api_experiment.go

```diff
@@ -477,3 +488,52 @@ func (a *apiServer) GetExperimentCheckpoints(
 		resp.Checkpoints, req.OrderBy, req.SortBy, apiv1.GetExperimentCheckpointsRequest_SORT_BY_TRIAL_ID)
 	return resp, a.paginate(&resp.Pagination, &resp.Checkpoints, req.Offset, req.Limit)
 }
+
+func (a *apiServer) CreateExperiment(
+	ctx context.Context, req *apiv1.CreateExperimentRequest,
+) (*apiv1.CreateExperimentResponse, error) {
+	detParams := CreateExperimentParams{
+		ConfigBytes:  protojson.Format(req.Config),
+		ModelDef:     filesToArchive(req.ModelDefinition),
+		ValidateOnly: req.ValidateOnly,
+	}
+	if req.ParentId != 0 {
+		parentID := int(req.ParentId)
+		detParams.ParentID = &parentID
+	}
+
+	dbExp, validateOnly, err := a.m.postParseExperiment(&detParams)
+
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid experiment: %s", err)
+	}
+
+	if validateOnly {
+		// old api: return nil, c.NoContent(http.StatusNoContent)
+		// or we can use codes.already exists.
+		// if we don't want to respond with an error we could
+		// 1. change the response to a union that has an OK
+		// 2. convert dbExp to the proto equivalent and respond with that.
+		// 3. return an empty or null experiment
+		// return nil, status.Errorf(codes.Aborted, "experiment is valid")
```

Comment:
> return nil, status.Errorf(codes.Aborted, "experiment is valid")

This is fine, assuming that maps to a 204. I think it's just a gRPC quirk that it's an "error" but as long as it returns 2XX it's fine. If someone is using the gRPC client, theyll just need to inspect the code.

Reply:
hmm no this responds with 409. if the error code is in 200 range the the gateway expected to return a valid experiment of nil

https://github.com/grpc-ecosystem/grpc-gateway/blob/master/third_party/googleapis/google/rpc/code.proto